ui-lib/src/main/java/co/electriccoin/zcash/ui/common/FlowExt.kt

```diff
@@ -0,0 +1,52 @@
+package co.electriccoin.zcash.ui.common
+
+import kotlinx.coroutines.Deferred
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.async
+import kotlinx.coroutines.coroutineScope
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.flow
+import kotlinx.coroutines.withContext
+import kotlin.time.Duration
+import kotlin.time.ExperimentalTime
+import kotlin.time.TimeSource
+
+@OptIn(ExperimentalTime::class)
+fun <T> Flow<T>.throttle(
+    duration: Duration,
+    timeSource: TimeSource = TimeSource.Monotonic
+): Flow<T> = flow {
+    coroutineScope {
+        val context = coroutineContext
+        var timeMark = timeSource.markNow()
+        var delayEmit: Deferred<Unit>? = null
+        var lastEmittedValue: T? = null
+
+        collect { value ->
+            if (lastEmittedValue == null) {
+                emit(value)
+                lastEmittedValue = value
+                timeMark = timeSource.markNow()
+                return@collect
+            }
+
+            if (timeMark.elapsedNow() >= duration) {
+                emit(value)
+                lastEmittedValue = value
+                delayEmit?.cancel()
+                timeMark = timeSource.markNow()
+            } else {
+                delayEmit?.cancel()
+                delayEmit = async(Dispatchers.IO) {
```

Comment:
I have a concern that we could potentially end up emitting an old value due to a race condition.

If the timing is right, I think a newly collected value in the outer collect could emit and then the async emit could occur with the prior value.

I think it would be helpful to see if we can write a test case that triggers this condition.  It would be a non-deterministic test case (e.g. doesn't fail every time), but if we run the test in a loop it should occur frequently enough that it becomes a robust verification.  Honza did something similar for the SDK's DatabaseCoordinatorTest.